### ISSUE TITLE
ValueAccessor.writeValue is being called twice, first time with a phantom null value

### ISSUE THREAD

**Issue by alexi2014** [angular/angular#14988]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
The ValueAccessor.writeValue method is called twice on a custom value accessor during component initialization, when a control is bound using [(ngModel)], first time with a phantom null value.

Note that this is not an issue in form builder scenarios when control is bound using [formControlName].

**Expected behavior**
ValueAccessor.writeValue should be called only once, with an actual value of the property bound to [(ngModel)].

**Minimal reproduction of the problem with instructions**
Run the following plunker:
[http://plnkr.co/edit/g5kjqQ9bQ9DF2cs4DbEH?p=preview](http://plnkr.co/edit/g5kjqQ9bQ9DF2cs4DbEH?p=preview)

MyValueAccessor is applied to the MyCmp component, the latter is bound using [(ngModel)] to the property initialized to 'Hello'. MyValueAccessor.writeValue writes the passed value to the console, the output looks as follows:
```
MyValueAccessor.writeValue: null
MyValueAccessor.writeValue: Hello
```

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

* **Angular version:** RC.4 - 2.4.9

* **Browser:** Chrome, IE11 

 
* **Language:** TypeScript

* **Node (for AoT issues):** `node --version` =

**Comment by alexzuza** [comment]
Reactive forms are synchronous. Template-driven forms are asynchronous. It's a difference that matters. 
And i think that is why you can see such behavior

https://angular.io/docs/ts/latest/guide/reactive-forms.html#!#async-vs-sync

**Comment by alexi2014** [comment]
I don't understand how this relates to synchronous or asynchronous. ngModel is bound to a property declared as
theValue = 'Hello';
, but receives a null on initialization. If theValue never has a null value, why null is passed to the accessor?

**Comment by alexi2014** [comment]
@IgorMinar Thanks Igor. When do you think it could be fixed?

**Comment by DzmitryShylovich** [comment]
> why null is passed to the accessor?

because `FormControl` is initialized with `null` value. Just add a check `if (obj != null) {}`

**Comment by alexi2014** [comment]
Yes, and this check will break initialization in reactive forms scenario, where this "phantom" null is not passed.

**Comment by DzmitryShylovich** [comment]
how it will break initialization in reactive forms scenario?

**Comment by alexi2014** [comment]
If initial value of a bound property is null, the `if (obj != null)` check you suggested will ignore this assignment and not propagate it to the control where ValueAccessor is applied to.

**Comment by DzmitryShylovich** [comment]
Ok I see how to fix it but they don't merge fixes until v4 release :)

**Comment by alexi2014** [comment]
@DzmitryShylovich  Why do you think it can't fixed in v2?

**Comment by DzmitryShylovich** [comment]
it will be fixed in 2.x but the fix will be available after v4 release

**Comment by alexi2014** [comment]
I see, thanks for the clarification. It's better then nothing anyway:)

**Comment by CSchulz** [comment]
It is the same in Angular 4.1.0:
http://plnkr.co/edit/cOHEHCu55X2MNEUxBhou?p=preview

**Comment by wilsoncook** [comment]
+1

**Comment by myagoo** [comment]
What is the status on this ?

**Comment by dpxxdp** [comment]
This is an open issue for us too.  @DzmitryShylovich, any status on this?

**Comment by anjmao** [comment]
Any news on this issue? I'm creating custom component and for my case it is very important to do some logic on first writeValue call, but since only second writeValue call is giving real value this is a big issue :)

**Comment by mofogasy** [comment]
Problem still exists in Angular 5.2.1.

Any news on that? Is this by design?

**Comment by jcimoch** [comment]
Still present in latest angular version.

**Comment by RoyiNamir** [comment]
I met this issue in Nativescript angular and asked here : https://stackoverflow.com/q/49615038/859154
solution for me was to wrap  inside the component which implementControlValueAccessor : with 
```
<ng-container *ngIf="myValInHere">
...
 </ng-container>
```

**Comment by SV-efi** [comment]
This is still an issue. 

I have a nested user control 
form
  <partslist>
        <part>

Form  component : 
<partsList [(ngModel)]="partsList">

PartsList Component :       
<div *ngFor="let p of parts;let i=index">
     <part name="part-{{i}}"  [(ngModel)]="parts[i]"></part>             
<div>

Part Component
 <div>{{part.partNumber}}</div>
 <input [(ngModel)]="part.quantity" type="number" />

When writeValue is first called with null, I can put a safety band aid around it to check for null.
But the template engine screws up the [(ngModel)]= "part.quantity" in the Part Component and complains about "Cannot read property 'quantity' of undefined".

**Comment by ckelley7** [comment]
This bug is ruining my marriage.  Please fix.  I know.  I'm not worthy.

**Comment by RoyiNamir** [comment]
@ckelley7  You can mitigate this bug by using 

```
 writeValue(value: any) {
     if (value!=null){
        this.innerValue = value;
    }
   }
```

Then  , even if it get null value at startup , you can ignore this call.

**Comment by RoyiNamir** [comment]
@ckelley7  You can check for null value 

http://plnkr.co/edit/UNYbyNK609F9o2vOtzRy?p=preview

```
 writeValue(obj: any): void {
    
    if (obj!=null){
    
    console.log(`MyValueAccessor.writeValue: ${obj}`);
    }
  }
```

Until it's fixed. At lest you won't get exception.

**Comment by SV-efi** [comment]
@RoyiNamir : Thanks for the suggestion, but this won't help the errors encountered if you reference your innerValue in your template. 

My component template has 
<input [(ngModel)]="innerValue.quantity"/>

The template engine still tries to navigate to innerValue.quantity with this error 
ERROR TypeError: Cannot read property 'quantity' of undefined

**Comment by RoyiNamir** [comment]
@SV-efi  Have you tried wrapping it with a Form tag? as this will cause initialization to be schedule : 

problematic : https://plnkr.co/edit/fOL57Hjw2PAOx6WXh36t?p=preview

Solution :  https://plnkr.co/edit/5Ez9VgN0yNz7QaF7DMck?p=preview

**Comment by SV-efi** [comment]
@RoyiNamir : Yes. I have a form tag in my outermost component. Having _another_ form tag inside my innermost component would mess up my UI workflow.

**Comment by RoyiNamir** [comment]
@SV-efi Good. Then remove that `form` and wrap with `div` instead with `ngForm` directive.

You won't convince me that another div mess up your UI workflow in a manner that can't be solved in 10 min.

```
 <div ngForm>          <----------
          <custom-input    name="someValue" 
                          [(ngModel)]="dataModel">
              Write in this wrapper control:
          </custom-input>
          </div>
```

**Comment by SV-efi** [comment]
Adding a   `<div ngForm>` gets  the template going after the 2nd time writeValue, after littering  the console window with  n errors where n => number of my inner components. But the ng-valid class does not bubble up. It is stopped at the div ngForm. 

![image](https://user-images.githubusercontent.com/23062859/41255316-f4772ec0-6d7a-11e8-9a0c-6fbfae159026.png)

**Comment by mloremi** [comment]
i tried all the solution here none of em work.
skipping the null value but the next model change doesnt seem to get reflected in the view

**Comment by nikhilgoud** [comment]
Mine is a different case I have to support null values too, I tried with ngModelChange with no luck

The issue I have is when the ngModel value is set in the OnInit of a component, the Parent writeValue receives the values but Child WriteValue is receiving null!
https://stackblitz.com/edit/angular-ysfuxj

Can this be fixed, any other recommended approaches?

**Comment by slavafomin** [comment]
Hello!

Thank you for your hard work, however, what is the status on this 1.5 years lasting issue?

Is this designated as a bug?

Thank you!

**Comment by vinagreti** [comment]
The problem is in v7 too. You can see a demo it here: https://stackblitz.com/edit/angular-hsernc.
I did not understand if it is the way it was designed to work or if it is a bug.

**Comment by larssn** [comment]
The best thing about this bug is if you're using 3rd party components that doesn't know about these issues.

It gives life flavour

**Comment by andrea78novi** [comment]
Is anyone from the Angular team aware that this bug is very harmful?

**Comment by daveyx** [comment]
i am just facing this bug too. if i init my custom formcontrol (ControlValueAccessor) with a value, it is passed when writeValue() is called the first time. writeValue() is called a second time with the phantom null value.
We are using Angular 7.2.0
I've expected writeValue() with a null value when the FormGroup was reset()

**Comment by geertjanb** [comment]
> Adding a `<div ngForm>` gets the template going after the 2nd time writeValue, after littering the console window with n errors where n => number of my inner components. But the ng-valid class does not bubble up. It is stopped at the div ngForm.
> 
> ![image](https://user-images.githubusercontent.com/23062859/41255316-f4772ec0-6d7a-11e8-9a0c-6fbfae159026.png)

the null check prevents wrong values being send. Now facing the issue as quoted above.

**Comment by RyanAntenor** [comment]
Any solutions on this issue?
I noticed also if you use it on *ngFor  it update all same custom element value to its latest.
Meaning if value was 1, 1 and 5 it update from a same custom element..isn't supposed to be encapsulated?

**Comment by jcimoch** [comment]
I just leave this comment as reminder to check in two years if this was fixed.

**Comment by sangnguyent1905** [comment]
this is my solution
private _handleValueChange(value: Date) {
    if (this.selectedValue || value) {
      this.writeValue(value);
      this.onChange(value);
    }
}

**Comment by vthinkxie** [comment]
Hi @IgorMinar 

this bug would trigger an extra dirty check cycle in most form component, should this issue add to v9.x-candidates?

**Comment by liorwohl** [comment]
Experiencing this bug now in Angular 9.0.4, I have this CSS in my UI since for ever, so users wont click submit of form without changes:
`.ng-pristine button[type=submit] {
  opacity: 0.5;
  pointer-events: none; 
}`
 I created a new component that implements ControlValueAccessor, now a form that is using this new component is not ng-pristine without the user touching anything. 
I cant even find a workaround.

**Comment by liorwohl** [comment]
Simple way to avoid this bug, in your writeValue function, instead of calling this.onChange(value) without checking anything, add originalValue property to the component and these conditions:

`writeValue (value) {
    if (value !== null || this.originalValue !== undefined) {
      if (this.originalValue !== undefined && value !== this.originalValue) {
        this.onChange(value);
      }
      this.originalValue = value;
    }
  }`

This way you will get .ng-dirty only when the user actually changes the value, not the phantom null and not on initially setting a default value for the ngModel.

**Comment by AnonymousArthur** [comment]
@liorwohl this only works for number or string. Things like array or object wouldn't work, becuase accurately doing A !== B could be very expensive.

**Comment by liorwohl** [comment]
@AnonymousArthur for objects and arrays I guess you can use JSON.stringify. If you are making an input replacement using ControlValueAccessor, stringifying two objects only when the user do some action probably has a very negligible performance hit.

**Comment by AnonymousArthur** [comment]
@liorwohl yes, JSON.stringify could be a workaround for array or object in most of time. However, you should still be careful of using it.

**Comment by whizkidwwe1217** [comment]
This bug takes longer to be fixed than creating a vaccine for Covid-19.

**Comment by paolomangiadev** [comment]
> This bug takes longer to be fixed than creating a vaccine for Covid-19.

imagine Covid-19 has been caused by this bug

**Comment by etay2000** [comment]
I'm surprised this hasn't been automatically locked yet.

In all fairness, this issue has only been open for 3 years. Would do you guys expect from a small company like Google?

**Comment by jcimoch** [comment]
> 
> 
> I'm surprised this hasn't been automatically locked yet.
> 
> In all fairness, this issue has only been open for 3 years. Would do you guys expect from a small company like Google?

Husshhh 🤐  Do not disturb the angular team, they are probably rewriting compiler again.

**Comment by belyan** [comment]
It's not funny. This defect is already 3 years old. Angular team, fix it ASAP.

**Comment by SV-efi** [comment]
Choose your pick. 
* This bug will be fixed in Angular v.17
* Another Javascript framework will replace Angular so this will be a moot issue.
* Covid will wipe out the entire earth, so there is no point fussing over this bug.

**Comment by N2D4** [comment]
#38140 fixes this. Hope your marriage is alright, @ckelley7 !

**Comment by shorbagyDH** [comment]
Until the bug fix is merged for people who still have this problem and the problem for calling `writeValue` when using `setValue` or `patchValue`.
 `ControlValueAccessor` extended component don't ever try to use `setValue` or `patchValue` in passed `formControl` always try to keep the changes of the `formControl` value through `onChange` method and if you have to listen to `valueChanges` for example and you `setValue` inside it try to have another `Subject` or `BehaviourSubject` to lock the `valueChanges` & `writeValue` if you are setting the value of the `FormControl`.

**Comment by mohamedelshorbagy** [comment]
Until the bug fix is merged for people who still have this problem and the problem for calling `writeValue` when using `setValue` or `patchValue`.
`ControlValueAccessor` extended component don't ever try to use `setValue` or `patchValue` in passed `formControl` always try to keep the changes of the `formControl` value through `onChange` method and if you have to listen to `valueChanges` for example and you `setValue` inside it try to have another `Subject` or `BehaviourSubject` to lock the `valueChanges` & `writeValue` if you are setting the value of the `FormControl`.

**Comment by Twynzen** [comment]
I come from the future, it's a bad year. This bug has not been fixed yet. They say to create a condition to detect if the "writeValue" is null.  But "writeValue" belongs to the nodeModules, therefore go to "@angular/forms" and modify it, it seems a very bad idea. That I understood. Explain to me if I'm wrong.  lo siento por mi ingles, hablo español

**Comment by ckelley7** [comment]
> I come from the future, it's a bad year. This bug has not been fixed yet. They say to create a condition to detect if the "writeValue" is null. But "writeValue" belongs to the nodeModules, therefore go to "@angular/forms" and modify it, it seems a very bad idea. That I understood. Explain to me if I'm wrong. Sorry for my english, i speak spanish.

Tu inglés es lo suficientemente bueno como para hacer reír a carcajadas a un hablante nativo, Daniel. Creo que lo que están sugiriendo es modificar el método ValueAccessor.writeValue de su control personalizado para detectar nulo y hacer lo necesario. Me he equivocado antes.

**Comment by nullifiedtsk** [comment]
Any progress on this?
Stuck with the same thing, the value produced once for non-standalone controls, but for standalone controls always emits `null` first. As workaround i am started to use `formControls` instead of `ngModel` bindings, but it just makes code complicated.

**Comment by kagan94** [comment]
I also faced with same issue.

Temporary workaround might be to run change detector after setting value in `writeValue`:   
```
  constructor(private changeDetection: ChangeDetectorRef) {
  }

  ...

  public writeValue(obj: any) {
    this.selected = obj;
    this.changeDetection.markForCheck();
  }
```

**Comment by jforjava1981** [comment]
https://stackoverflow.com/questions/39835811/ngmodel-binding-is-null-in-oninit-in-angular-2
@N2D4 is this issue related to question asked in above link?
If not Is there a way I can somehow initialise the value in my custom component using ngcontrolvalueaccessor with [(ngModel)]?

**Comment by afusa** [comment]
This is getting ridiculous now. When is this getting fixed ?

**Comment by SadeemGxG** [comment]
So is there any progress?

**Comment by maxime1992** [comment]
As an image is worth a 1000 words, hopefully that'll make more sense than trying to explain any further

| .                                                                                                              | .                                                                                                              | .                                                                                                              |
| -------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- |
| ![image](https://user-images.githubusercontent.com/4950209/130599501-b41c5d85-da8f-4610-85bb-0b041250a0d0.png) | ![image](https://user-images.githubusercontent.com/4950209/130599585-f86e400d-ca8d-4333-8ccd-6c5e819c4d8a.png) | ![image](https://user-images.githubusercontent.com/4950209/130599608-8e772d3a-0724-4f86-b389-02ed0f34e640.png) |
| ![image](https://user-images.githubusercontent.com/4950209/130599625-7d93e51b-5b7f-4377-a9cb-3bc82ecdecd8.png) | ![image](https://user-images.githubusercontent.com/4950209/130599557-36a2f99e-a3a4-428e-acde-9ca2b076fd05.png) | ![image](https://user-images.githubusercontent.com/4950209/130600137-8f19a5db-5931-44f9-9aed-12a82bdda770.png)                                                                                                              |

**Comment by stewx** [comment]
@maxime1992 There is already an open PR, wise guy. https://github.com/angular/angular/pull/38140

**Comment by luketanner** [comment]
> @maxime1992 There is already an open PR, wise guy. It's been open for over a year: #38140

@stewx Well, even-wiser-guy, now that PR is **closed** and this bug remains **unfixed** after **FIVE years**.

**Comment by Expertus777** [comment]
How to fix it simple and properly?

**Comment by martynas9** [comment]
In my case. I've noticed that first `writeValue` with a `null` value is called before `registerOnChange` is called, and then, the `writeValue` is called with the real value again.

So my workaround for this bug was this:

Create class property for onChange but don't assign a value to it:
```typescript
onChange: (value: unknown) => void;
```

Then on `registerOnChange` we assign a function to onChange property:
```typescript
  registerOnChange(fn: (value: unknown) => void): void {
    this.onChange = fn;
  }
```

And check if `onChange` is set in writeValue function:
```typescript
writeValue(value: unknown): void {
    if(this.onChange) {
      // your code here
    }
  }
  ```
  
  Maybe it will be helpful for someone.

**Comment by william-fargues** [comment]
Hey Angular team, I've had time to start a family, see my children grow up, graduate and learn React instead of Angular just to avoid this bug, do you think this is normal?

**Comment by amitbeck** [comment]
Wasn't this partially solved with [`nonNullable`](https://angular.io/api/forms/FormControlOptions#nonNullable) (which replaced the deprecated [`initialValueIsDefault`](https://angular.io/api/forms/FormControlOptions#initialValueIsDefault)) property of [`FormControlOptions`](https://angular.io/api/forms/FormControlOptions)?

**Comment by sergey-dudik** [comment]
any updates here ?

**Comment by thuanduongsts** [comment]
what the fuck it last long till 2023 without any solution ?

**Comment by raysuelzer** [comment]
I really regret using angular. I've noticed that a bunch of my form controls and bindings are now broken because it seems like they changed the order the lifecycle callbacks in v15. (ngModelChange) gets called before it's even passed a value now.

**Comment by yuriybelike** [comment]
👆 Funny how this bites even Angular contributors.

Maybe it's time to fix it?

**Comment by raysuelzer** [comment]
> 👆 Funny how this bites even Angular contributors.
> 
> Maybe it's time to fix it?

The problem is that this has been a bug for so long the team let it become a feature. Any angular app has had to handle this "phantom null" value for so long now that actually fixing the underlying issue would probably break a bunch of stuff since everyone now knows to check:
```
writeValue(data: any) {
  // Angular bug: first writeValue is always null. 
  // hopefully we never really need null here. 
  if (data == null) { return; }
}
```

**Comment by raysuelzer** [comment]
> In my case. I've noticed that first `writeValue` with a `null` value is called before `registerOnChange` is called, and then, the `writeValue` is called with the real value again.
> 
> So my workaround for this bug was this:
> 
> Create class property for onChange but don't assign a value to it:
> 
> ```ts
> onChange: (value: unknown) => void;
> ```
> 
> Then on `registerOnChange` we assign a function to onChange property:
> 
> ```ts
>   registerOnChange(fn: (value: unknown) => void): void {
>     this.onChange = fn;
>   }
> ```
> 
> And check if `onChange` is set in writeValue function:
> 
> ```ts
> writeValue(value: unknown): void {
>     if(this.onChange) {
>       // your code here
>     }
>   }
> ```
> 
> Maybe it will be helpful for someone.

This is a top tier hack.

**Comment by Almost-Infinity** [comment]
Hello! I have another case that related to this issue.

I have a range slider with two `FormControl` that can be either number or null. The default value of those controls is `null`. So, during the initialization Form API is trying to call `writeValue` with the `null`. Because a slider (slider thumb component) [is not initialized yet](https://github.com/angular/components/blob/3500908ca609aa4d2110513ea20fbaff475b78a9/src/material/slider/slider-input.ts#L538), the initial value does not reach the slider (thumb).

Then, when I want to reset the `FormControl` (by calling `FormControl.reset()`) and the control [is already initialized](https://github.com/angular/components/blob/3500908ca609aa4d2110513ea20fbaff475b78a9/src/material/slider/slider-input.ts#L550) [this](https://github.com/angular/components/blob/3500908ca609aa4d2110513ea20fbaff475b78a9/src/material/slider/slider-input.ts#L538) condition returns true and my value reach the slider (thumb). So, I have unexpected behavior (for me).

Briefly:
1. During initialization, the default `null` does not reach the slider;
2. If I call `FormControl.reset()` after the initialization, then the default `null` does;

Q: Is this correct and expected behavior?

**NG:** 17.0.6
**Node:** v18.17.1
**Browser:** 120.0.6099.109
**Platform:** macOS 14.2 (23C64)

<details>
  <summary>Piece of code</summary>
  
  ```typescript
  @Component({
    template: `
      <mat-slider>
        <input matSliderStartThumb [formControl]="startControl" />
        <input matSliderEndThumb [formControl]="endControl" />
      </mat-slider>

      <button (click)="onResetClick()">Reset</button>
    `,
  })
  export class MyComponent {
    readonly startControl = new FormControl<number | null>(null);
    readonly endControl = new FormControl<numner | null(null);
  
    onResetClick(): void {
      this.startControl.reset();
      this.endControl.reset();
    }
  }
  ```
</details>

**Comment by howlettga** [comment]
Just thought I would throw in the scenario of using a directive on top of the control. Because the control initially renders with a null value, the directive only has access to this null value during ngAfterViewInit lifecycle method. Despite a real value being initially supplied to the control, the init lifecycle methods aren't able to correctly interpret the initial value.
Not sure there is a workaround within the lifecycle method; would be nice if this very old bug would be addressed.

**Comment by sonu-spotdraft** [comment]
Manually updating the value for the host element works for me.
```typescript

constructor(private elementRef: ElementRef) {}
 writeValue(email: string): void {
    this.inputValue = email;
    /**
     * Manually updating the element value
     */
    this.ref.nativeElement.value = this.inputValue;
  }
```

**Comment by shajz** [comment]
This hack (https://github.com/angular/angular/issues/14988#issuecomment-1310420293) works **only** if your CVA is used with `[(ngModel)]`. If the CVA is used with `[formControl]` / `formControlName` then you have a new problem. The first `writeValue` is not a phantom one so you'll miss it. **That's very bad.**

You need to check if the current `ngControl` is an `ngModel` but if you provide `NG_VALUE_ACCESSOR` (like in a usual CVA), you can't just `inject(NgControl)`: you'll get a circular dependency in DI.

So you have to do this:

```ts
@Component({
	// ...
	providers: [
		{
			provide: NG_VALUE_ACCESSOR,
			useExisting: forwardRef(() => YourCVA),
			multi: true,
		},
	],
})
export class YourCVA implements OnInit, ControlValueAccessor {
	#injector = inject(Injector);
	#ngControl: NgControl; // Initialized in ngOnInit

	#onChange: (value: unknown) => void;

	// ...

	ngOnInit(): void {
		this.#ngControl = this.#injector.get(NgControl);
	}

	registerOnChange(fn: (value: unknown) => void): void {
		this.#onChange = fn;
	}

	// ...

	writeValue(date: unknown): void {
		if (this.#ngControl instanceof NgModel && !this.#onChange) {
			// avoid phantom call for ngModel
			// https://github.com/angular/angular/issues/14988#issuecomment-2946355465
			return;
		}

		// your code
	}
}
```
